### PR TITLE
Add support for updating replicas.

### PR DESCRIPTION
**Pull request by atoaca**

### CODE REVIEW

File: coriolis/db/api.py [hunk 1]
```diff
@@ -543,3 +543,17 @@ def add_task_progress_update(context, task_id, current_step, total_steps,
...
+@enginefacade.writer
+def update_replica(context, replica_id, updated_values):
+    replica = get_replica(context, replica_id)
+    if not replica:
+        raise exception.NotFound("Replica not found")
+    for n in ["source_environment", "destination_environment", "notes",
+              "network_map", "storage_mappings"]:
+        if n in updated_values:
+            setattr(replica, n, updated_values[n])
+    # the oslo_db library uses this method for both the `created_at` and
```
[the remark this answers is not in the document]
Comment (atoaca) [reply coriolis/db/api.py]: the recommended way to get timestamps in sqlalchemy is with `func.now()` but oslo_db uses `timeutils.utcnow()` for both `created_at` and `updated_at` fields(which are inherited in our tables) hence my justification for consistency's sake.

File: coriolis/providers/base.py [hunk 2]
```diff
@@ -387,3 +387,27 @@ class BaseEndpointStorageProvider(object, with_metaclass(abc.ABCMeta)):
...
+class BaseReplicaUpdateableProvider(object, with_metaclass(abc.ABCMeta)):
+    """ Class for replica providers (export and import) which offer the
+    functionality of updating the parameters for a replica.
+    """
+    @abc.abstractmethod
+    def check_update_environment_params(
+            self, ctxt, connection_info, export_info, volumes_info, old_params,
+            new_params):
+        """ Checks that any existing replica resources for the VM given by its
+        `export_info` which were replicating using the `old_params` is
```
Comment (aznashwan) [inline coriolis/providers/base.py]: s/replicating/replicated (my bad sorry)

File: coriolis/api/v1/replicas.py [hunk 3]
```diff
@@ -109,6 +110,150 @@ def delete(self, req, id):
...
+        if final_storage_mappings:
+            final_values[
+                'destination_environment'][
+                'storage_mappings'] = final_storage_mappings
+        if final_network_map:
+            final_values[
+                'destination_environment'][
+                'network_map'] = final_network_map
+
+        new_instances = updated_values.get('instances', [])
```
[the remark this answers is not in the document]
Comment (atoaca) [reply coriolis/api/v1/replicas.py]: Yes, I did, updating instances works, but I agree that it opens up a can of worms so to speak, so i'll be deleting this field.

File: coriolis/api/v1/replicas.py [hunk 4]
```diff
@@ -109,6 +110,135 @@ def delete(self, req, id):
...
+            storage_mappings['default'] = default_storage_backend
+
+        return storage_mappings
+
+    def _get_merged_replica_values(self, replica, updated_values):
+        final_values = {}
+        # NOTE: this just replaces options at the top-level and does not do
+        # merging of container types (ex: lists, dicts)
+        for option in [
+            "source_environment", "destination_environment", "network_map"]:
```
Comment (aznashwan) [inline coriolis/api/v1/replicas.py]: This line needs another indent.

File: coriolis/conductor/rpc/server.py [hunk 5]
```diff
@@ -861,6 +861,9 @@ def _handle_post_task_actions(self, ctxt, task, execution, task_info):
                 LOG.debug(
                     "No 'transfer_result' was returned for task type '%s' "
                     "for transfer action '%s'", task_type, execution.action_id)
+        elif task_type == constants.TASK_TYPE_UPDATE_REPLICA:
+            # NOTE perform the actual db update
```
Comment (aznashwan) [inline coriolis/conductor/rpc/server.py]: Nit: `# NOTE: ` (note (pun not intended) the colon)

File: coriolis/db/api.py [hunk 6]
```diff
@@ -543,3 +543,18 @@ def add_task_progress_update(context, task_id, current_step, total_steps,
...
     task_progress_update.message = message
+
+
+@enginefacade.writer
+def update_replica(context, replica_id, updated_values):
+    replica = get_replica(context, replica_id)
+    if not replica:
+        raise exception.NotFound("Replica not found")
+    for n in ["source_environment", "destination_environment", "notes",
+              "network_map", "storage_mappings", "instances"]:
```
Comment (aznashwan) [inline coriolis/db/api.py]: Please remove `instances` from this list.

File: coriolis/api/v1/replicas.py [hunk 7]
```diff
@@ -109,6 +110,135 @@ def delete(self, req, id):
...
+        if final_network_map:
+            final_values[
+                'destination_environment'][
+                'network_map'] = final_network_map
+
+        return final_values
+
+    def _validate_update_body(self, id, context, body):
+
+        replica = self._replica_api.get_replica(context, id)
```
Comment (aznashwan) [inline coriolis/api/v1/replicas.py]: Please add explicit checks and return an explicit error message is someone tries to update `origin_endpoint_id`, `destination_endpoint_id` or `instances`.

**Pull request merged**